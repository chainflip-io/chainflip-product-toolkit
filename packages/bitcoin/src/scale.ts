import {
  assetConstants,
  type ChainflipAsset,
  type ChainflipChain,
} from '@chainflip/utils/chainflip';
import { Struct, Bytes, u16, u128, u8, Vector } from 'scale-ts';

const addressByteLengths: Record<ChainflipChain, number | undefined> = {
  Bitcoin: undefined,
  Arbitrum: 20,
  Ethereum: 20,
  Solana: 32,
  Polkadot: 32,
  Assethub: 32,
};

export const createSwapDataCodec = (asset: ChainflipAsset) =>
  Struct({
    version: u8,
    destinationAsset: u8,
    destinationAddress: Bytes(addressByteLengths[assetConstants[asset].chain]),
    parameters: Struct({
      retryDuration: u16,
      minOutputAmount: u128,
      numberOfChunks: u16,
      chunkInterval: u16,
      boostFee: u8,
      brokerFee: u8,
      affiliates: Vector(Struct({ accountIndex: u8, commissionBps: u8 })),
    }),
  });
