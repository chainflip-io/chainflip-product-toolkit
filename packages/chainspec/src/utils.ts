import assert from 'assert';
import * as prettier from 'prettier';
import { type ParsedMetadata } from './BaseParser';

type ChangeType = 'added' | 'removed' | 'changed';
type Change = { path: string[]; type: ChangeType };

/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
// a function to deep diff two metadata objects
function* diff(a: any, b: any, path: string[] = []): Generator<Change> {
  const addedKeys = Object.keys(b)
    .filter((key) => b[key] !== undefined)
    .filter((key) => !a[key]);
  const removedKeys = Object.keys(a)
    .filter((key) => a[key] !== undefined)
    .filter((key) => !b[key]);
  const commonKeys = Object.keys(a)
    .filter((key) => a[key] !== undefined)
    .filter((key) => b[key]);

  for (const key of addedKeys) {
    yield { path: [...path, key], type: 'added' };
  }

  for (const key of removedKeys) {
    yield { path: [...path, key], type: 'removed' };
  }

  for (const key of commonKeys) {
    if (typeof a[key] === 'object' && typeof b[key] === 'object') {
      yield* diff(a[key], b[key], [...path, key]);
    } else if (a[key] !== b[key]) {
      yield { path: [...path, key], type: 'changed' };
    }
  }
}
/* eslint-enable @typescript-eslint/no-unsafe-member-access */
/* eslint-enable @typescript-eslint/no-unsafe-argument */

class ChangelogGenerator {
  addedOrRemovedPallets: Record<string, 'added' | 'removed'> = {};

  changedEvents: Record<
    `${string}.${string}`,
    { field: string; type: 'added' | 'changed' | 'removed' }[]
  > = {};
  addedOrRemovedEvents: Record<`${string}.${string}`, 'added' | 'removed'> = {};

  trackedFieldChanges = new Set<string>();

  add({ path, type }: Change) {
    if (path.length === 1) {
      assert(type === 'added' || type === 'removed');
      this.addedOrRemovedPallets[path[0]] = type;
    } else if (path.length === 2) {
      const name = `${path[0]}.${path[1]}` as const;
      assert(type === 'added' || type === 'removed');
      this.addedOrRemovedEvents[name] = type;
    } else {
      const [pallet, event, , field] = path;
      const id = `${pallet}.${event}.${field}`;
      if (this.trackedFieldChanges.has(id)) return;
      this.trackedFieldChanges.add(id);
      (this.changedEvents[`${pallet}.${event}`] ??= []).push({ field, type });
    }
  }

  toString() {
    const lines: string[] = [];

    const pallets = Object.keys(this.addedOrRemovedPallets).sort();

    if (pallets.length) {
      lines.push('New or removed pallets:');
      for (const pallet of pallets) {
        lines.push(`  ${pallet}: ${this.addedOrRemovedPallets[pallet]}`);
      }
      lines.push('');
    }

    const events = Object.keys(this.addedOrRemovedEvents)
      .concat(Object.keys(this.changedEvents))
      .sort() as `${string}.${string}`[];

    let currentPallet = '';
    for (const event of events) {
      const [pallet, name] = event.split('.');
      if (pallet !== currentPallet) {
        if (currentPallet) lines.push('');
        currentPallet = pallet;
        lines.push(`${pallet}:`);
      }
      if (event in this.addedOrRemovedEvents) {
        lines.push(`  - ${name}: ${this.addedOrRemovedEvents[event]}`);
      } else {
        lines.push(`  - ${name}:`);
        const fields = this.changedEvents[event].map(
          ({ field, type }) => `    - ${field}: ${type}`,
        );
        lines.push(...fields);
      }
    }

    lines.push('');
    return lines.join('\n');
  }
}

export const diffSpecs = (a: ParsedMetadata, b: ParsedMetadata) => {
  const changedOrAddedEvents = new Set<string>();
  const changelog = new ChangelogGenerator();

  for (const change of diff(a, b)) {
    changelog.add(change);

    const [pallet, event, field] = change.path;

    if (event && (field || change.type === 'added')) {
      // if a field changed or a new event was added
      changedOrAddedEvents.add(`${pallet}.${event}`);
    } else if (change.type === 'added') {
      // if a new pallet was added
      Object.keys(b[pallet])
        .map((e) => `${pallet}.${e}`)
        .forEach((ev) => changedOrAddedEvents.add(ev), changedOrAddedEvents);
    }
  }

  return { changedOrAddedEvents, changelog: changelog.toString() };
};

export const formatCode = (code: string) =>
  prettier.format(code, {
    parser: 'typescript',
    singleQuote: true,
    trailingComma: 'all',
    printWidth: 100,
  });
