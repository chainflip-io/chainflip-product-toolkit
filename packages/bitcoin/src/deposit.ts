import { assert } from '@chainflip/utils/assertion';
import * as base58 from '@chainflip/utils/base58';
import { bytesToHex } from '@chainflip/utils/bytes';
import { type ChainflipAsset, assetConstants, assetContractId } from '@chainflip/utils/chainflip';
import { POLKADOT_SS58_PREFIX } from '@chainflip/utils/consts';
import * as ss58 from '@chainflip/utils/ss58';
import { type VaultSwapData } from '@chainflip/utils/types';
import BigNumber from 'bignumber.js';
import * as rpc from './rpc';
import { createSwapDataCodecV0, createSwapDataCodecV1, UtxoDataV0, UtxoDataV1 } from './scale';

const encodeChainAddress = (data: Uint8Array, asset: ChainflipAsset) => {
  switch (assetConstants[asset].chain) {
    case 'Solana':
      return base58.encode(data);
    case 'Assethub':
    case 'Polkadot':
      return ss58.encode({ data, ss58Format: POLKADOT_SS58_PREFIX });
    case 'Ethereum':
    case 'Arbitrum':
      return bytesToHex(data);
    case 'Bitcoin':
      return new TextDecoder().decode(data);
  }
};

const contractIdToInternalAsset: Record<number, ChainflipAsset> = Object.fromEntries(
  (Object.entries(assetContractId) as [ChainflipAsset, number][]).map(([asset, id]) => [id, asset]),
);

const parseVaultSwapData = (data: Uint8Array) => {
  const version = data[0];
  const contractId = data[1];
  const outputAsset = contractIdToInternalAsset[contractId];

  let destinationAddress: Uint8Array;
  let parameters: UtxoDataV0 | UtxoDataV1;

  if (version === 1) {
    ({ destinationAddress, parameters } = createSwapDataCodecV1(outputAsset).dec(data));
  } else if (version === 0) {
    ({ destinationAddress, parameters } = createSwapDataCodecV0(outputAsset).dec(data));
  } else {
    throw new Error('unsupported version');
  }

  assert(outputAsset, 'unknown asset contract id');

  return {
    ...parameters,
    outputAsset,
    destinationAddress: encodeChainAddress(destinationAddress, outputAsset),
  };
};

const getX128PriceFromAmounts = (depositAmount: bigint, minOutputAmount: bigint) =>
  BigInt(
    new BigNumber(minOutputAmount.toString())
      .div(depositAmount.toString())
      .multipliedBy(new BigNumber(2).pow(128))
      .toFixed(0, BigNumber.ROUND_FLOOR),
  );

export type BitcoinVaultSwapData = VaultSwapData<null> & { depositAddress: string };

export const findVaultSwapData = async (
  url: string,
  txId: string,
): Promise<BitcoinVaultSwapData | null> => {
  const tx = await rpc.makeRequest(url, 'getrawtransaction', [txId, true]);

  if (!tx) return null;

  const data = parseVaultSwapData(tx.vout[1].scriptPubKey.hex);

  const amount = tx.vout[0].value;

  const block = tx.blockhash
    ? await rpc.makeRequest(url, 'getblock', [tx.blockhash, true]).catch(() => null)
    : null;

  return {
    inputAsset: 'Btc' as const,
    amount,
    depositAddress: tx.vout[0].scriptPubKey.address,
    refundParams: {
      refundAddress: tx.vout[2].scriptPubKey.address,
      retryDuration: data.retryDuration,
      minPrice: getX128PriceFromAmounts(amount, data.minOutputAmount),
      maxOraclePriceSlippage: 'maxOraclePriceSlippage' in data ? data.maxOraclePriceSlippage : null,
    },
    destinationAddress: data.destinationAddress,
    outputAsset: data.outputAsset,
    brokerFee: { account: null, commissionBps: data.brokerFee },
    affiliateFees: data.affiliates,
    ccmDepositMetadata: null,
    maxBoostFee: data.boostFee,
    dcaParams:
      data.numberOfChunks === 1 && data.chunkInterval === 2
        ? null
        : { chunkInterval: data.chunkInterval, numberOfChunks: data.numberOfChunks },
    depositChainBlockHeight: block && block.height,
  };
};
