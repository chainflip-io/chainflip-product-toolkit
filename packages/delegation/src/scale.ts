import { unreachable } from '@chainflip/utils/assertion';
import { bytesToHex } from '@chainflip/utils/bytes';
import * as ss58 from '@chainflip/utils/ss58';
import { _void, Bytes, CodecType, Enum, Option, Struct, u128 } from 'scale-ts';

const DelegationApi = Enum({
  Delegate: Struct({
    operator: Bytes(32),
  }),
  Undelegate: _void,
  SetMaxBid: Struct({
    maybeMaxBid: Option(u128),
  }),
});

export const EthereumSCApi = Enum({
  Delegation: DelegationApi,
});

type CallData =
  | { call: 'delegate'; operator: `cF${string}` }
  | { call: 'undelegate' }
  | {
      call: 'setMaxBid';
      /** The new max bid, if set to `undefined`, it will be */
      maxBid?: bigint | undefined;
    };

const buildDelegationApiData = (data: CallData): CodecType<typeof DelegationApi> => {
  switch (data.call) {
    case 'delegate':
      return {
        tag: 'Delegate',
        value: { operator: ss58.decode(data.operator).data },
      } as const;
    case 'undelegate':
      return { tag: 'Undelegate', value: undefined } as const;
    case 'setMaxBid':
      return { tag: 'SetMaxBid', value: { maybeMaxBid: data.maxBid } } as const;
    default:
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      return unreachable(data, `Unsupported call type: ${(data as any)?.type}`);
  }
};

export const serializeDelegationCall = (data: CallData): `0x${string}` =>
  bytesToHex(EthereumSCApi.enc({ tag: 'Delegation', value: buildDelegationApiData(data) }));
