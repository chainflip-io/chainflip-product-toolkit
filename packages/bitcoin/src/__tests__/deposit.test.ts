import * as base58 from '@chainflip/utils/base58';
import { hexToBytes } from '@chainflip/utils/bytes';
import { assetConstants, assetContractId, type ChainflipAsset } from '@chainflip/utils/chainflip';
import * as ss58 from '@chainflip/utils/ss58';
import { describe, expect, it, type MockInstance } from 'vitest';
import { spyOn } from '@/testing';
import { findVaultSwapData } from '../deposit';
import { createSwapDataCodec } from '../scale';
import { tx, block } from './fixtures';

const mockFetch = (results: unknown[], status = 200) =>
  results.reduce(
    (mock: MockInstance<typeof fetch>, result) =>
      mock.mockResolvedValueOnce({
        ok: status === 200,
        status,
        text: () => Promise.resolve(JSON.stringify(result)),
      } as Response),
    spyOn(globalThis, 'fetch'),
  );

const buildNullData = ({
  destinationAsset,
  destinationAddress,
  chunkInterval = 1,
  numberOfChunks = 2,
  retryDuration = 100,
  minOutputAmount = BigInt(10 ** assetConstants[destinationAsset].decimals),
  boostFee = 0,
  brokerFee = 0,
  affiliates = [],
}: {
  destinationAsset: ChainflipAsset;
  destinationAddress: Uint8Array;
  chunkInterval?: number;
  numberOfChunks?: number;
  retryDuration?: number;
  minOutputAmount?: bigint;
  boostFee?: number;
  brokerFee?: number;
  affiliates?: { accountIndex: number; commissionBps: number }[];
}) => {
  const codec = createSwapDataCodec(destinationAsset);

  const bytes = codec.enc({
    version: 0,
    destinationAsset: assetContractId[destinationAsset],
    destinationAddress,
    parameters: {
      retryDuration,
      affiliates,
      boostFee,
      brokerFee,
      chunkInterval,
      numberOfChunks,
      minOutputAmount,
    },
  });

  return [
    0x6a, // OP_RETURN
    bytes.length, // OP_PUSHBYTES_XX
    ...bytes,
  ]
    .map((n) => n.toString(16).padStart(2, '0'))
    .join('');
};

const addresses = {
  Eth: hexToBytes('0xa56a6be23b6cf39d9448ff6e897c29c41c8fbdff'),
  ArbEth: hexToBytes('0xa56a6be23b6cf39d9448ff6e897c29c41c8fbdff'),
  Sol: base58.decode('A8dScaeGChTgzvok95uVAxd44LRkq5ckcpppNVQwkLQb'),
  Dot: ss58.decode('167ZvRqc7V6HrEUSvtV8c3JRUtPjJhHXUpAwhmPktAGj1uzq').data,
  Btc: new TextEncoder().encode('tb1qhjurnfz4qah4rg7ntue6x287ehdvded20rj9vh'),
  HubDot: ss58.decode('167ZvRqc7V6HrEUSvtV8c3JRUtPjJhHXUpAwhmPktAGj1uzq').data,
};

type Mutable<T> = T extends readonly (infer U)[]
  ? Mutable<U>[]
  : T extends Record<string | number | symbol, any>
    ? { -readonly [K in keyof T]: Mutable<T[K]> }
    : T;

describe(findVaultSwapData, () => {
  const address1 = 'tb1qvwmuc3pjhwju287sjs5vg7467t2jlymn35lwsj';
  const address2 = 'tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6';

  it.each([
    ['Eth', 1, 2, 30, 5, 1, 100, address1, undefined],
    ['ArbEth', 1, 2, 15, 10, 0.1, 90, address2, undefined],
    ['Sol', 20, 100, 10, 15, 0.01, 80, undefined, undefined],
    ['Dot', 1, 2, 0, 0, 0.001, 69, undefined, [{ accountIndex: 1, commissionBps: 10 }]],
    ['HubDot', 1, 2, 0, 0, 0.001, 69, undefined, [{ accountIndex: 1, commissionBps: 10 }]],
    ['Btc', 1, 2, 5, 20, 0.0001, 50, undefined, [{ accountIndex: 2, commissionBps: 20 }]],
  ] as const)(
    'gets the vault swap data (%s)',
    async (
      destinationAsset,
      numberOfChunks,
      chunkInterval,
      boostFee,
      brokerFee,
      depositAmount,
      retryDuration,
      refundAddress,
      affiliates,
    ) => {
      const nulldata = buildNullData({
        destinationAsset,
        destinationAddress: addresses[destinationAsset],
        chunkInterval,
        numberOfChunks,
        boostFee,
        brokerFee,
        retryDuration,
        affiliates: affiliates as Mutable<typeof affiliates>,
      });

      mockFetch([tx({ nulldata, depositAmount, refundAddress }), block]);

      mockFetch([{ result: { height: 3902357 }, error: null, id: 1 }]);

      const data = await findVaultSwapData(
        'https://bitcoin.rpc',
        '77a4dcda118d8cd4e537616effeac741ff60dbdb7af0b7f2f54a3a15c0556239',
      );

      expect(data).toMatchSnapshot();
    },
  );

  it('returns null if the tx is not found', async () => {
    mockFetch([
      {
        result: null,
        error: {
          code: -5,
          message:
            'No such mempool or blockchain transaction. Use gettransaction for wallet transactions.',
        },
        id: 1,
      },
    ]);

    const data = await findVaultSwapData(
      'https://bitcoin.rpc',
      '77a4dcda118d8cd4e537616effeac741ff60dbdb7af0b7f2f54a3a15c0556239',
    );

    expect(data).toBeNull();
  });

  it('returns null if the tx is not a vault swap', async () => {
    const txInfo = tx({ nulldata: '00', depositAmount: 1 });
    txInfo.result.vout.splice(1);
    mockFetch([txInfo]);

    const data = await findVaultSwapData(
      'https://bitcoin.rpc',
      '77a4dcda118d8cd4e537616effeac741ff60dbdb7af0b7f2f54a3a15c0556239',
    );

    expect(data).toBeNull();
  });

  it('throws parse errors', async () => {
    mockFetch([tx({ nulldata: '00', depositAmount: 1 }), block]);

    mockFetch([{ result: { height: 3902357 }, error: null, id: 1 }]);

    await expect(
      findVaultSwapData(
        'https://bitcoin.rpc',
        '77a4dcda118d8cd4e537616effeac741ff60dbdb7af0b7f2f54a3a15c0556239',
      ),
    ).rejects.toThrowErrorMatchingInlineSnapshot(`[Error: unsupported version]`);
  });

  it('throws rpc errors', async () => {
    spyOn(globalThis, 'fetch').mockResolvedValueOnce({
      ok: true,
      status: 200,
      text: () =>
        Promise.resolve(
          JSON.stringify({
            result: null,
            error: { code: -32601, message: 'Method not found' },
            id: 1,
          }),
        ),
    } as Response);

    await expect(
      findVaultSwapData(
        'https://bitcoin.rpc',
        '77a4dcda118d8cd4e537616effeac741ff60dbdb7af0b7f2f54a3a15c0556239',
      ),
    ).rejects.toThrowErrorMatchingInlineSnapshot(`[Error: RPC error [-32601]: Method not found]`);
  });

  it('returns null for block if the request fails', async () => {
    const nulldata = buildNullData({
      destinationAsset: 'Eth',
      destinationAddress: addresses.Eth,
    });

    mockFetch([
      tx({ nulldata, depositAmount: 1 }),
      { id: 1, jsonrpc: '2.0', result: {}, error: null },
    ]);

    expect(
      await findVaultSwapData(
        'https://bitcoin.rpc',
        '77a4dcda118d8cd4e537616effeac741ff60dbdb7af0b7f2f54a3a15c0556239',
      ),
    ).toMatchSnapshot();
  });

  it.each([null, undefined])('returns null for block if no hash found', async (blockHash) => {
    const nulldata = buildNullData({
      destinationAsset: 'Eth',
      destinationAddress: addresses.Eth,
    });

    const txInfo = tx({ nulldata, depositAmount: 1 });

    // @ts-expect-error -- for testing
    txInfo.result.blockhash = blockHash;

    mockFetch([txInfo, { id: 1, jsonrpc: '2.0', result: {}, error: null }]);

    expect(
      await findVaultSwapData(
        'https://bitcoin.rpc',
        '77a4dcda118d8cd4e537616effeac741ff60dbdb7af0b7f2f54a3a15c0556239',
      ),
    ).toMatchSnapshot();
  });

  it('returns null for block if no hash found', async () => {
    const nulldata = buildNullData({
      destinationAsset: 'Eth',
      destinationAddress: addresses.Eth,
    });

    const txInfo = tx({ nulldata, depositAmount: 1 });

    // @ts-expect-error -- for testing
    delete txInfo.result.blockhash;

    mockFetch([txInfo, { id: 1, jsonrpc: '2.0', result: {}, error: null }]);

    expect(
      await findVaultSwapData(
        'https://bitcoin.rpc',
        '77a4dcda118d8cd4e537616effeac741ff60dbdb7af0b7f2f54a3a15c0556239',
      ),
    ).toMatchSnapshot();
  });

  it("returns null for block if the block isn't found", async () => {
    const nulldata = buildNullData({
      destinationAsset: 'Eth',
      destinationAddress: addresses.Eth,
    });

    mockFetch([tx({ nulldata, depositAmount: 1 })]);
    mockFetch(
      [{ id: 1, jsonrpc: '2.0', result: null, error: { code: -5, message: 'Block not found' } }],
      500,
    );

    spyOn(globalThis, 'fetch').mockRejectedValueOnce(new Error('Failed to fetch'));

    expect(
      await findVaultSwapData(
        'https://bitcoin.rpc',
        '77a4dcda118d8cd4e537616effeac741ff60dbdb7af0b7f2f54a3a15c0556239',
      ),
    ).toMatchSnapshot();
  });

  it('throws parse errors', async () => {
    mockFetch([{ id: 1, jsonrpc: '2.0', result: {} }]);

    await expect(
      findVaultSwapData(
        'https://bitcoin.rpc',
        '77a4dcda118d8cd4e537616effeac741ff60dbdb7af0b7f2f54a3a15c0556239',
      ),
    ).rejects.toThrowErrorMatchingInlineSnapshot(`
      [ZodError: [
        {
          "code": "invalid_union",
          "unionErrors": [
            {
              "issues": [
                {
                  "code": "invalid_type",
                  "expected": "null",
                  "received": "object",
                  "path": [
                    "result"
                  ],
                  "message": "Expected null, received object"
                },
                {
                  "code": "invalid_type",
                  "expected": "object",
                  "received": "undefined",
                  "path": [
                    "error"
                  ],
                  "message": "Required"
                }
              ],
              "name": "ZodError"
            },
            {
              "issues": [
                {
                  "code": "invalid_type",
                  "expected": "null",
                  "received": "undefined",
                  "path": [
                    "error"
                  ],
                  "message": "Required"
                }
              ],
              "name": "ZodError"
            }
          ],
          "path": [],
          "message": "Invalid input"
        }
      ]]
    `);
  });
});
