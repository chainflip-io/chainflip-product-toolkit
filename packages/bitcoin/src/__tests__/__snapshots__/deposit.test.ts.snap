// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`findVaultSwapData > gets the vault swap data (ArbEth) 1`] = `
{
  "affiliateFees": [],
  "amount": 10000000n,
  "brokerFee": {
    "account": null,
    "commissionBps": 10,
  },
  "ccmDepositMetadata": null,
  "dcaParams": null,
  "depositAddress": "tb1pjpqkt5ghr3lts922ykufct49trhmkjesmvmrfrq88w7fd0ewwtaq0da2gg",
  "depositChainBlockHeight": 3902357,
  "destinationAddress": "0xa56a6be23b6cf39d9448ff6e897c29c41c8fbdff",
  "inputAsset": "Btc",
  "maxBoostFee": 15,
  "outputAsset": "ArbEth",
  "refundParams": {
    "maxOraclePriceSlippage": null,
    "minPrice": 34028236692093846346337460743176821145600000000000n,
    "refundAddress": "tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6",
    "retryDuration": 90,
  },
}
`;

exports[`findVaultSwapData > gets the vault swap data (Btc) 1`] = `
{
  "affiliateFees": [
    {
      "accountIndex": 2,
      "commissionBps": 20,
    },
  ],
  "amount": 10000n,
  "brokerFee": {
    "account": null,
    "commissionBps": 20,
  },
  "ccmDepositMetadata": null,
  "dcaParams": null,
  "depositAddress": "tb1pjpqkt5ghr3lts922ykufct49trhmkjesmvmrfrq88w7fd0ewwtaq0da2gg",
  "depositChainBlockHeight": 3902357,
  "destinationAddress": "tb1qhjurnfz4qah4rg7ntue6x287ehdvded20rj9vh",
  "inputAsset": "Btc",
  "maxBoostFee": 5,
  "outputAsset": "Btc",
  "refundParams": {
    "maxOraclePriceSlippage": 5,
    "minPrice": 3402823669209384634633746074317682114560000n,
    "refundAddress": "tb1qhjurnfz4qah4rg7ntue6x287ehdvded20rj9vh",
    "retryDuration": 50,
  },
}
`;

exports[`findVaultSwapData > gets the vault swap data (Dot) 1`] = `
{
  "affiliateFees": [
    {
      "accountIndex": 1,
      "commissionBps": 10,
    },
  ],
  "amount": 100000n,
  "brokerFee": {
    "account": null,
    "commissionBps": 0,
  },
  "ccmDepositMetadata": null,
  "dcaParams": null,
  "depositAddress": "tb1pjpqkt5ghr3lts922ykufct49trhmkjesmvmrfrq88w7fd0ewwtaq0da2gg",
  "depositChainBlockHeight": 3902357,
  "destinationAddress": "167ZvRqc7V6HrEUSvtV8c3JRUtPjJhHXUpAwhmPktAGj1uzq",
  "inputAsset": "Btc",
  "maxBoostFee": 0,
  "outputAsset": "Dot",
  "refundParams": {
    "maxOraclePriceSlippage": null,
    "minPrice": 34028236692093846346337460743176821145600000n,
    "refundAddress": "tb1qhjurnfz4qah4rg7ntue6x287ehdvded20rj9vh",
    "retryDuration": 69,
  },
}
`;

exports[`findVaultSwapData > gets the vault swap data (Eth) 1`] = `
{
  "affiliateFees": [],
  "amount": 100000000n,
  "brokerFee": {
    "account": null,
    "commissionBps": 5,
  },
  "ccmDepositMetadata": null,
  "dcaParams": null,
  "depositAddress": "tb1pjpqkt5ghr3lts922ykufct49trhmkjesmvmrfrq88w7fd0ewwtaq0da2gg",
  "depositChainBlockHeight": 3902357,
  "destinationAddress": "0xa56a6be23b6cf39d9448ff6e897c29c41c8fbdff",
  "inputAsset": "Btc",
  "maxBoostFee": 30,
  "outputAsset": "Eth",
  "refundParams": {
    "maxOraclePriceSlippage": null,
    "minPrice": 3402823669209384634633746074317682114560000000000n,
    "refundAddress": "tb1qvwmuc3pjhwju287sjs5vg7467t2jlymn35lwsj",
    "retryDuration": 100,
  },
}
`;

exports[`findVaultSwapData > gets the vault swap data (HubDot) 1`] = `
{
  "affiliateFees": [
    {
      "accountIndex": 1,
      "commissionBps": 10,
    },
  ],
  "amount": 100000n,
  "brokerFee": {
    "account": null,
    "commissionBps": 0,
  },
  "ccmDepositMetadata": null,
  "dcaParams": null,
  "depositAddress": "tb1pjpqkt5ghr3lts922ykufct49trhmkjesmvmrfrq88w7fd0ewwtaq0da2gg",
  "depositChainBlockHeight": 3902357,
  "destinationAddress": "167ZvRqc7V6HrEUSvtV8c3JRUtPjJhHXUpAwhmPktAGj1uzq",
  "inputAsset": "Btc",
  "maxBoostFee": 0,
  "outputAsset": "HubDot",
  "refundParams": {
    "maxOraclePriceSlippage": 1,
    "minPrice": 34028236692093846346337460743176821145600000n,
    "refundAddress": "tb1qhjurnfz4qah4rg7ntue6x287ehdvded20rj9vh",
    "retryDuration": 69,
  },
}
`;

exports[`findVaultSwapData > gets the vault swap data (Sol) 1`] = `
{
  "affiliateFees": [],
  "amount": 1000000n,
  "brokerFee": {
    "account": null,
    "commissionBps": 15,
  },
  "ccmDepositMetadata": null,
  "dcaParams": {
    "chunkInterval": 100,
    "numberOfChunks": 20,
  },
  "depositAddress": "tb1pjpqkt5ghr3lts922ykufct49trhmkjesmvmrfrq88w7fd0ewwtaq0da2gg",
  "depositChainBlockHeight": 3902357,
  "destinationAddress": "A8dScaeGChTgzvok95uVAxd44LRkq5ckcpppNVQwkLQb",
  "inputAsset": "Btc",
  "maxBoostFee": 10,
  "outputAsset": "Sol",
  "refundParams": {
    "maxOraclePriceSlippage": null,
    "minPrice": 340282366920938463463374607431768211456000n,
    "refundAddress": "tb1qhjurnfz4qah4rg7ntue6x287ehdvded20rj9vh",
    "retryDuration": 80,
  },
}
`;

exports[`findVaultSwapData > returns null for block if no hash found 1`] = `
{
  "affiliateFees": [],
  "amount": 100000000n,
  "brokerFee": {
    "account": null,
    "commissionBps": 0,
  },
  "ccmDepositMetadata": null,
  "dcaParams": {
    "chunkInterval": 1,
    "numberOfChunks": 2,
  },
  "depositAddress": "tb1pjpqkt5ghr3lts922ykufct49trhmkjesmvmrfrq88w7fd0ewwtaq0da2gg",
  "depositChainBlockHeight": null,
  "destinationAddress": "0xa56a6be23b6cf39d9448ff6e897c29c41c8fbdff",
  "inputAsset": "Btc",
  "maxBoostFee": 0,
  "outputAsset": "Eth",
  "refundParams": {
    "maxOraclePriceSlippage": null,
    "minPrice": 3402823669209384634633746074317682114560000000000n,
    "refundAddress": "tb1qhjurnfz4qah4rg7ntue6x287ehdvded20rj9vh",
    "retryDuration": 100,
  },
}
`;

exports[`findVaultSwapData > returns null for block if no hash found 2`] = `
{
  "affiliateFees": [],
  "amount": 100000000n,
  "brokerFee": {
    "account": null,
    "commissionBps": 0,
  },
  "ccmDepositMetadata": null,
  "dcaParams": {
    "chunkInterval": 1,
    "numberOfChunks": 2,
  },
  "depositAddress": "tb1pjpqkt5ghr3lts922ykufct49trhmkjesmvmrfrq88w7fd0ewwtaq0da2gg",
  "depositChainBlockHeight": null,
  "destinationAddress": "0xa56a6be23b6cf39d9448ff6e897c29c41c8fbdff",
  "inputAsset": "Btc",
  "maxBoostFee": 0,
  "outputAsset": "Eth",
  "refundParams": {
    "maxOraclePriceSlippage": null,
    "minPrice": 3402823669209384634633746074317682114560000000000n,
    "refundAddress": "tb1qhjurnfz4qah4rg7ntue6x287ehdvded20rj9vh",
    "retryDuration": 100,
  },
}
`;

exports[`findVaultSwapData > returns null for block if no hash found 3`] = `
{
  "affiliateFees": [],
  "amount": 100000000n,
  "brokerFee": {
    "account": null,
    "commissionBps": 0,
  },
  "ccmDepositMetadata": null,
  "dcaParams": {
    "chunkInterval": 1,
    "numberOfChunks": 2,
  },
  "depositAddress": "tb1pjpqkt5ghr3lts922ykufct49trhmkjesmvmrfrq88w7fd0ewwtaq0da2gg",
  "depositChainBlockHeight": null,
  "destinationAddress": "0xa56a6be23b6cf39d9448ff6e897c29c41c8fbdff",
  "inputAsset": "Btc",
  "maxBoostFee": 0,
  "outputAsset": "Eth",
  "refundParams": {
    "maxOraclePriceSlippage": null,
    "minPrice": 3402823669209384634633746074317682114560000000000n,
    "refundAddress": "tb1qhjurnfz4qah4rg7ntue6x287ehdvded20rj9vh",
    "retryDuration": 100,
  },
}
`;

exports[`findVaultSwapData > returns null for block if the block isn't found 1`] = `
{
  "affiliateFees": [],
  "amount": 100000000n,
  "brokerFee": {
    "account": null,
    "commissionBps": 0,
  },
  "ccmDepositMetadata": null,
  "dcaParams": {
    "chunkInterval": 1,
    "numberOfChunks": 2,
  },
  "depositAddress": "tb1pjpqkt5ghr3lts922ykufct49trhmkjesmvmrfrq88w7fd0ewwtaq0da2gg",
  "depositChainBlockHeight": null,
  "destinationAddress": "0xa56a6be23b6cf39d9448ff6e897c29c41c8fbdff",
  "inputAsset": "Btc",
  "maxBoostFee": 0,
  "outputAsset": "Eth",
  "refundParams": {
    "maxOraclePriceSlippage": null,
    "minPrice": 3402823669209384634633746074317682114560000000000n,
    "refundAddress": "tb1qhjurnfz4qah4rg7ntue6x287ehdvded20rj9vh",
    "retryDuration": 100,
  },
}
`;

exports[`findVaultSwapData > returns null for block if the request fails 1`] = `
{
  "affiliateFees": [],
  "amount": 100000000n,
  "brokerFee": {
    "account": null,
    "commissionBps": 0,
  },
  "ccmDepositMetadata": null,
  "dcaParams": {
    "chunkInterval": 1,
    "numberOfChunks": 2,
  },
  "depositAddress": "tb1pjpqkt5ghr3lts922ykufct49trhmkjesmvmrfrq88w7fd0ewwtaq0da2gg",
  "depositChainBlockHeight": null,
  "destinationAddress": "0xa56a6be23b6cf39d9448ff6e897c29c41c8fbdff",
  "inputAsset": "Btc",
  "maxBoostFee": 0,
  "outputAsset": "Eth",
  "refundParams": {
    "maxOraclePriceSlippage": null,
    "minPrice": 3402823669209384634633746074317682114560000000000n,
    "refundAddress": "tb1qhjurnfz4qah4rg7ntue6x287ehdvded20rj9vh",
    "retryDuration": 100,
  },
}
`;
